simd: add simd_size_type

ldh4 opened the pull request:
Related to #5674 and https://github.com/kokkos/kokkos/pull/6109#discussion_r1275432904.

Introduced an alias, `simd_size_type`, for a signed int type.
Replaced size_t and int uses by `simd_size_type` as described in [p1928r6](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p1928r6.pdf).

nliber commented:
[#6109 (comment)](https://github.com/kokkos/kokkos/pull/6109#discussion_r1275432904) was asking for a `simd_size_type` to be a private type.  This change makes it a public type.  Is that intentional?

I guess I'm asking whether or not this should be in `Kokkos::Impl` (or equivalent) or not.

I don't know that the answer is obvious one way or the other.  For instance, `simd_mask<T>::size()` right now returns it, which would argue in favor of it being public.  (If it were private, then `simd_mask<T>` ought to have a typedef for it.)

ldh4 commented:
> [#6109 (comment)](https://github.com/kokkos/kokkos/pull/6109#discussion_r1275432904) was asking for a `simd_size_type` to be a private type. This change makes it a public type. Is that intentional?
> 
> I guess I'm asking whether or not this should be in `Kokkos::Impl` (or equivalent) or not.
> 
> I don't know that the answer is obvious one way or the other. For instance, `simd_mask<T>::size()` right now returns it, which would argue in favor of it being public. (If it were private, then `simd_mask<T>` ought to have a typedef for it.)

It was also unclear to me whether placing `simd_size_type` in an impl namespace provides a value. As you described, there will need an alias for `Impl::simd_size_type` to be used publicly in all backends. (Or `using simd_size_type = Impl::internal_simd_size_type`, if we want to preserve the names as consistently as possible with the synopsis provided in [p1928r7](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p1928r7.pdf) section 7.1.2)

diehlpk reviewed (dismissed):
LGTM! However, we should check for compilation and tests.

nmm0 commented on simd/src/Kokkos_SIMD_AVX512.hpp:
```diff
@@ -2111,8 +2091,8 @@ class basic_simd<std::int64_t, simd_abi::avx512_fixed_size<8>> {
-  operator[](std::size_t i) const {
-    value_type tmp[size()];
+  operator[](simd_size_t i) const {
+    value_type tmp[size];
```
Doesn't matter too much but with integral constant you can still use `size()` if you want. Not necessary to change it back though

ldh4 replied:
I did initially change them all to be just `size` that uses the conversion operator and changed them back to `size()` for readability, but it looks like I missed few of these while reverting them.
Reverted remaining of these to be `size()` for consistency.

JBludau commented on simd/src/Kokkos_SIMD_AVX2.hpp:
```diff
@@ -1862,9 +1838,7 @@ class basic_simd<std::int64_t, simd_abi::avx2_fixed_size<4>> {
-  KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION static constexpr std::size_t size() {
-    return 4;
-  }
+  static constexpr std::integral_constant<simd_size_t, 4> size{};
```
I hope this does not break something downstream, but if it does they were doing stuff that was brittle...

nmm0 replied:
It is experimental, so no guarantee things won't change -- but also we have to do it this way if we want to be as close to standard as possible

JBludau reviewed (approved):
This looks good and I think we can afford merging it so close to the release

dalg24 commented on simd/src/Kokkos_SIMD_Common_Math.hpp:
```diff
@@ -45,7 +45,7 @@ KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION T
   auto const& m = x.impl_get_mask();
   auto result   = Kokkos::reduction_identity<T>::min();
-  for (std::size_t i = 0; i < v.size(); ++i) {
+  for (auto i = 0; i < v.size(); ++i) {
```
I am not sure `auto` is a good idea here.
It makes `i`  be an `int` so we type more and rely on size() returning a signed integer type.

ldh4 replied:
Converted them to `simd_size_t` so that they stay consistent with what `size()` will return.

dalg24 reviewed (commented):
Reiterating here that `simd-size-t` is exposition-only in the standard https://eel.is/c++draft/simd#expos.defn-1 which would typically mean `Impl::` for us.
I am ok with having it public if you add a comment in the code at the definition and a note in the documentation.

You could add a compile test that size() returns simd_size_t and that it is a signed integral type.

ldh4 commented:
> Reiterating here that `simd-size-t` is exposition-only in the standard https://eel.is/c++draft/simd#expos.defn-1 which would typically mean `Impl::` for us. I am ok with having it public if you add a comment in the code at the definition and a note in the documentation.
> 
> You could add a compile test that size() returns simd_size_t and that it is a signed integral type.

Placed `simd_size_t` into `Impl` namespace.

dalg24 commented on simd/src/Kokkos_SIMD.hpp:
Why are we qualifying with Experimental:: in this file?

ldh4 replied:
It's because most of this file is in `simd_abi::Impl` namespace, whereas `simd_size_t` is in `Experimental::Impl`.

dalg24 replied:
Thanks

dalg24 commented on simd/src/Kokkos_SIMD_AVX2.hpp:
```diff
@@ -44,7 +44,7 @@ namespace Experimental {
 namespace simd_abi {
 
-template <int N>
+template <Impl::simd_size_t N>
```
I don't understand why you would not need the Experimental:: qualifier here then

ldh4 replied:
I suppose the difference is being in the `simd_abi::Impl` namespace vs just the `simd_abi` namespace.
For instance, if I write `Impl::simd_size_t` (and not qualify `Experimental`) in `Kokkos_SIMD.hpp`, the compiler attempts to find the symbol `simd_abi::Impl::simd_size_t`. It fails and errors out.
```
/fgs/donlee/kokkos/kokkos/simd/src/Kokkos_SIMD.hpp:94:11: error: 'Kokkos::Experimental::simd_abi::Impl::simd_size_t' has not been declared
   94 | template <Impl::simd_size_t N>
```

dalg24 commented:
Seen in the CUDA-11.6-NVCC-DEBUG build on ORNL Jenkins
```
/var/jenkins/workspace/Kokkos_PR-6535/simd/unit_tests/include/TestSIMD_MaskOps.hpp:75:29: error: comparison of integer expressions of different signedness: 'std::size_t' {aka 'long unsigned int'} and 'Kokkos::Impl::integral_constant<int, 1>::value_type' {aka 'int'} [-Werror=sign-compare]
   75 |     for (std::size_t i = 0; i < mask_type::size(); ++i) {
      |                           ~~^~~~~~~~~~~~~~~~~~~
```

ldh4 commented:
> Seen in the CUDA-11.6-NVCC-DEBUG build on ORNL Jenkins
> 
> ```
> /var/jenkins/workspace/Kokkos_PR-6535/simd/unit_tests/include/TestSIMD_MaskOps.hpp:75:29: error: comparison of integer expressions of different signedness: 'std::size_t' {aka 'long unsigned int'} and 'Kokkos::Impl::integral_constant<int, 1>::value_type' {aka 'int'} [-Werror=sign-compare]
>    75 |     for (std::size_t i = 0; i < mask_type::size(); ++i) {
>       |                           ~~^~~~~~~~~~~~~~~~~~~
> ```

Fixed.

dalg24 commented:
Same CUDA-11.6-NVCC-DEBUG build still has
```
/var/jenkins/workspace/Kokkos_PR-6535/simd/unit_tests/include/SIMDTesting_Utilities.hpp:109:7: error: comparison of integer expressions of different signedness: 'std::size_t' {aka 'long unsigned int'} and 'Kokkos::Impl::integral_constant<int, 1>::value_type' {aka 'int'} [-Werror=sign-compare]
  109 |     if (n < result.size()) return false;
[...]
/var/jenkins/workspace/Kokkos_PR-6535/simd/unit_tests/include/SIMDTesting_Utilities.hpp:128:7: error: comparison of integer expressions of different signedness: 'std::size_t' {aka 'long unsigned int'} and 'Kokkos::Impl::integral_constant<int, 1>::value_type' {aka 'int'} [-Werror=sign-compare]
  128 |     if (n < result.size()) return false;

ldh4 commented:
> Same CUDA-11.6-NVCC-DEBUG build still has
> 
> ```
> /var/jenkins/workspace/Kokkos_PR-6535/simd/unit_tests/include/SIMDTesting_Utilities.hpp:109:7: error: comparison of integer expressions of different signedness: 'std::size_t' {aka 'long unsigned int'} and 'Kokkos::Impl::integral_constant<int, 1>::value_type' {aka 'int'} [-Werror=sign-compare]
>   109 |     if (n < result.size()) return false;
> [...]
> /var/jenkins/workspace/Kokkos_PR-6535/simd/unit_tests/include/SIMDTesting_Utilities.hpp:128:7: error: comparison of integer expressions of different signedness: 'std::size_t' {aka 'long unsigned int'} and 'Kokkos::Impl::integral_constant<int, 1>::value_type' {aka 'int'} [-Werror=sign-compare]
>   128 |     if (n < result.size()) return false;
> ```

That's odd. Locally, I'm somehow not getting those warning/errors even with `-Wall -Werror=sign-compare`, but those are valid errors that the CI has reported.
Regardless, fixed these errors along with other potential malformed comparisons in tests.

dalg24 commented:
:(
```
/var/jenkins/workspace/Kokkos_PR-6535/simd/src/Kokkos_SIMD_Common.hpp:400:29: error: comparison of integer expressions of different signedness: 'std::size_t' {aka 'long unsigned int'} and 'std::integral_constant<int, 8>::value_type' {aka 'int'} [-Werror=sign-compare]
  400 |   for (std::size_t i = 1; i < x.size(); ++i) {

dalg24 commented:
The 32-bit build on GH actions gives
```
[ RUN      ] simd.device_math_ops
SIMD unit test equality condition failed on device
Backtrace:
[0xf689ac5c] Kokkos::Impl::save_stacktrace()
[0xf686d0bf] Kokkos::Impl::host_abort(char const*)
[0x60813bbb] 
[0x608176bb] 
[0x60817e68] 
[0x608097e6] 
[0xf68151a2] GOMP_parallel
[0x608093f5] 
[0x6080a384] 
[0xf694e2da] void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)
[0xf693d7c7] testing::Test::Run()
[0xf693d9a9] testing::TestInfo::Run()
[0xf693db15] testing::TestSuite::Run()
[0xf693f8da] testing::internal::UnitTestImpl::RunAllTests()
[0xf694e91a] bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)
[0xf693dc9a] testing::UnitTest::Run()
[0x607fa945] 
[0xf6246c75] 
[0xf6246d38] __libc_start_main
[0x607fb7b7]

ldh4 commented:
> The 32-bit build on GH actions gives
> 
> ```
> [ RUN      ] simd.device_math_ops
> SIMD unit test equality condition failed on device
> Backtrace:
> [0xf689ac5c] Kokkos::Impl::save_stacktrace()
> [0xf686d0bf] Kokkos::Impl::host_abort(char const*)
> [0x60813bbb] 
> [0x608176bb] 
> [0x60817e68] 
> [0x608097e6] 
> [0xf68151a2] GOMP_parallel
> [0x608093f5] 
> [0x6080a384] 
> [0xf694e2da] void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)
> [0xf693d7c7] testing::Test::Run()
> [0xf693d9a9] testing::TestInfo::Run()
> [0xf693db15] testing::TestSuite::Run()
> [0xf693f8da] testing::internal::UnitTestImpl::RunAllTests()
> [0xf694e91a] bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)
> [0xf693dc9a] testing::UnitTest::Run()
> [0x607fa945] 
> [0xf6246c75] 
> [0xf6246d38] __libc_start_main
> [0x607fb7b7] 
> ```

This is fixed in #7960. It should go away once this PR is rebased after #7960 is merged.

crtrott commented:
So the other guy needs to go first.

ldh4 commented:
> So the other guy needs to go first.

Yes, but a correction. It's #8085 that fixed the issue, not #7960. #8085 has been merged. So the issue should be fixed now.
Rebased.

ldh4 commented:
Retest this please.

JBludau commented:
seeing (https://github.com/kokkos/kokkos/actions/runs/17840384806/job/50729772008?pr=6535, 32-bit build)
```
50/52 Test #50: Kokkos_UnitTest_SIMD .......................................Subprocess aborted***Exception:   0.17 sec
Kokkos::OpenMP::initialize WARNING: OMP_PROC_BIND environment variable not set
  In general, for best performance with OpenMP 4.0 or better set OMP_PROC_BIND=spread and OMP_PLACES=threads
  For best performance with OpenMP 3.1 set OMP_PROC_BIND=true
  For unit testing set OMP_PROC_BIND=false

[==========] Running 18 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 18 tests from simd
[ RUN      ] simd.host_math_ops
[       OK ] simd.host_math_ops (0 ms)
[ RUN      ] simd.device_math_ops
SIMD unit test equality condition failed on device
Backtrace:
[0xf7b8fadc] Kokkos::Impl::save_stacktrace()
[0xf7b5e16f] Kokkos::Impl::host_abort(char const*)
[0x64a65b62] 
[0x64a67b8b] 
[0x64a68338] 
[0x64a54696] 
[0xf7b061a2] GOMP_parallel
[0x64a542ab] 
[0x64a55074] 
[0xf7c4242a] void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)
[0xf7c345d7] testing::Test::Run()
[0xf7c347b9] testing::TestInfo::Run()
[0xf7c34925] testing::TestSuite::Run()
[0xf7c34ffa] testing::internal::UnitTestImpl::RunAllTests()
[0xf7c42a6a] bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)
[0xf7c3527a] testing::UnitTest::Run()
[0x64a43c25] 
[0xf7537c75] 
[0xf7537d38] __libc_start_main
[0x64a44a67] 
```

dalg24 commented:
Plz resolve conflicts

ldh4 commented:
> Plz resolve conflicts

Squashed commits and rebased.
But I suspect that the error https://github.com/kokkos/kokkos/pull/6535#issuecomment-3312730794 still exists, despite having been handled in another merged PR. If it persists, I'll revisit this a bit later. I'm thinking that another change that I'm working on could allow bypassing the issue.

dalg24 commented:
The SNL-CI / SNL_CUDA_NVCC_12_2_CPP20 build fails with
```
/home/runner/_work/kokkos/kokkos/kokkos/simd/src/Kokkos_SIMD_Common.hpp: In instantiation of 'bool Kokkos::Experimental::any_of(const Kokkos::Experimental::basic_simd_mask<T, Abi>&) [with T = int; Abi = Kokkos::Experimental::simd_abi::avx512_fixed_size<16>]':
/home/runner/_work/kokkos/kokkos/kokkos/simd/src/Kokkos_SIMD_Common.hpp:416:15:   required from 'bool Kokkos::Experimental::none_of(const Kokkos::Experimental::basic_simd_mask<T, Abi>&) [with T = int; Abi = Kokkos::Experimental::simd_abi::avx512_fixed_size<16>]'
/home/runner/_work/kokkos/kokkos/kokkos/simd/src/Kokkos_SIMD_AVX512.hpp:4835:12:   required from here
/home/runner/_work/kokkos/kokkos/kokkos/simd/src/Kokkos_SIMD_Common.hpp:407:24: error: comparison of integer expressions of different signedness: 'size_t' {aka 'long unsigned int'} and 'std::integral_constant<int, 16>::value_type' {aka 'int'} [-Werror=sign-compare]
  407 |   for (size_t i = 0; i < basic_simd_mask<T, Abi>::size(); ++i) {
      |                      ~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

masterleinad commented:
Needs rebasing.

dalg24 commented:
It looks like the 32-bit build is still not passing

ldh4 commented on simd/unit_tests/include/TestSIMD_MathOps.hpp:
```diff
@@ -553,6 +553,13 @@ TEST(simd, host_math_ops) {
 }
 
 TEST(simd, device_math_ops) {
+#if defined(KOKKOS_IMPL_32BIT)
+  // FIXME_GCC: Floating point comparisons fail in a 32-bit build
+  // See https://github.com/kokkos/kokkos/pull/7912#issuecomment-2747713935
+  GTEST_SKIP()
+      << "skipping due to a GCC bug associated with the computation of "
+         "floating-point values in a 32-bit build.";
+#endif
```
Skipping the device math unit test for 32 bit build.
We've attempted several workarounds to manage this issue for the gcc 32 bit build (i.e. #7912 and #8085). However, this issue is recurring, and as described by @tretre91 in https://github.com/kokkos/kokkos/pull/7912#issuecomment-2747713935, the source of this issue is not in our implementation.

tretre91 replied:
It's a good thing we finally get rid of this, and since the 32 bit CI is host only, the math functions are still covered by the host test

tretre91 commented on simd/src/Kokkos_SIMD_AVX2.hpp:
You should replace `std::size_t` by `Impl::simd_size_t` in the `scatter_to` functions

ldh4 replied:
Applied.

tretre91 commented on simd/src/Kokkos_SIMD_NEON.hpp:
You should use `Impl::simd_size_t` in the `copy_from` and `gather_from` methods of `where_expression`

ldh4 replied:
I've been avoiding making changes in deprecated code. These are deprecated because they aren't part of std simd, and I'm not sure if their signatures should be updated along with rest of running code.

Now of course, `copy_from`, `gather_from` and `scatter_to` are all reintroduced into std simd (i.e. `[partial|unchecked]_load/store`), and when they are implemented in kokkos simd, `Impl::simd_size_t` will be used to follow their signatures in std simd.

tretre91 replied:
Ok, makes sense

tretre91 commented on simd/src/Kokkos_SIMD_SVE.hpp:
You should use `Impl::simd_size_t` in the shift operators taking an `int` argument

ldh4 replied:
Fixed. Forgot that SVE was merged after this PR was made.

ldh4 reviewed (commented):
> Other places where `Impl::simd_size_t` should be used:
> 
> * TestSIMD_Construction.hpp: the `expected_size` variable in `host_test_simd_default_abi`
> * TestSIMD_LoadStore.hpp: lines 44, 63 and 165
> * TestSIMD_BitwiseOps.hpp: It seems that this file has not been updated yet

Applied.

tretre91 commented on simd/src/Kokkos_SIMD_SVE.hpp:
```diff
@@ -384,7 +382,7 @@ class sve_mask<Derived, 32> {
           G, value_type,                                                       \
           std::integral_constant<                                              \
-              std::size_t,                                                     \
+              simd_size_t,                                                     \
```
```suggestion
              Impl::simd_size_t,                                               \
```

tretre91 commented on simd/src/Kokkos_SIMD_SVE.hpp:
```diff
@@ -412,9 +410,8 @@ class basic_simd<double, simd_abi::sve_fixed_size<SVE_DOUBLES_IN_VECTOR>> {
-    return SVE_DOUBLES_IN_VECTOR;
-  }
+  static constexpr std::integral_constant<simd_size_t, SVE_DOUBLES_IN_VECTOR>
+      size{};
```
```suggestion
  static constexpr std::integral_constant<Impl::simd_size_t,
                                          SVE_DOUBLES_IN_VECTOR>
      size{};
```

masterleinad commented:
https://cloud1.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-6535/23/pipeline-overview/ failed

dalg24 commented:
Uggg
Thanks for reporting
